### PR TITLE
Add learning rate cutoffs and automatic reduction

### PR DESCRIPTION
**Pull request by gelisam**
From Google Jules

### CODE REVIEW

**Comment by gelisam** [comment]
Before this, I was quite confused because small seemingly-inconsequential changes like changing the sample size had a large effect: the network seemed to want to learn the function `f(x) = 0` instead of `parity`, and the loss was oscillating around 0.5 :thinking:

I now think that I was just lucky that the network was learning at all before those changes, and that those small changes nudge the demo out of that lucky spot. I blame the zigzag detection code which automatically reduces the learning rate when the loss function starts to oscillate instead of continuously going down. When I am lucky, the zigzag detection kicks in once the network has found a pretty good starting point, and when I am unlucky, it kicks in at a poor location and learning stalls.

I don't actually want to improve the learning process so that it learns better, as this demo is not about training a neural network which always succeeds at minimizing the loss, but about spotting whether the model still satisfies its safety property once the model has converged. I do want the model to make a few mistakes once it has converged, so that there is something interesting to detect.

My solution is to let the network expore the space relatively randomly, at a "high temperature" (a learning rate of 0.3) until we detect that it is in a good spot: that is, until it has reached a low enough loss. This branch worked very well in that regard, in that it has allowed me to interactively figure out a concrete number for this "low enough" loss.

Now that I have found the number, I will just hardcode it in #38. So I will not merge this branch.
